File: sample/src/main/java/me/xiaopan/android/spear/sample/fragment/SearchFragment.java
package me.xiaopan.android.spear.sample.fragment;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.app.ActionBar;
import android.support.v7.app.ActionBarActivity;
import android.support.v7.widget.SearchView;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.ListView;
import android.widget.Toast;

import org.apache.http.HttpResponse;

import java.util.ArrayList;
import java.util.List;

import me.xiaoapn.android.spear.sample.R;
import me.xiaopan.android.gohttp.GoHttp;
import me.xiaopan.android.gohttp.HttpRequest;
import me.xiaopan.android.gohttp.HttpRequestFuture;
import me.xiaopan.android.gohttp.JsonHttpResponseHandler;
import me.xiaopan.android.inject.InjectContentView;
import me.xiaopan.android.inject.InjectExtra;
import me.xiaopan.android.inject.InjectView;
import me.xiaopan.android.inject.app.InjectFragment;
import me.xiaopan.android.spear.sample.activity.DetailActivity;
import me.xiaopan.android.spear.sample.adapter.SearchImageAdapter;
import me.xiaopan.android.spear.sample.net.request.SearchImageRequest;
import me.xiaopan.android.spear.sample.net.request.StarImageRequest;
import me.xiaopan.android.spear.sample.util.ScrollingPauseLoadManager;
import me.xiaopan.android.spear.sample.widget.HintView;
import me.xiaopan.android.widget.PullRefreshLayout;

/**
 * 图片搜索Fragment
 */
@InjectContentView(R.layout.fragment_search)
public class SearchFragment extends InjectFragment implements SearchImageAdapter.OnItemClickListener, PullRefreshLayout.OnRefreshListener {
    public static final String PARAM_OPTIONAL_STRING_SEARCH_KEYWORD = "PARAM_OPTIONAL_STRING_SEARCH_KEYWORD";

    @InjectView(R.id.refreshLayout_search) PullRefreshLayout pullRefreshLayout;
    @InjectView(R.id.list_search) private ListView recyclerView;
    @InjectView(R.id.hintView_search) private HintView hintView;

    private SearchImageRequest searchImageRequest;
    private HttpRequestFuture refreshRequestFuture;
    private SearchImageAdapter searchImageListAdapter;
    private MyLoadMoreListener loadMoreListener;

    @InjectExtra(PARAM_OPTIONAL_STRING_SEARCH_KEYWORD) private String searchKeyword = "美女";

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        searchImageRequest = new SearchImageRequest(searchKeyword);
        loadMoreListener = new MyLoadMoreListener();
        setHasOptionsMenu(true);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        setTitle(searchKeyword);
    }

    private void setTitle(String subtitle){
        if(getActivity() != null && getActivity() instanceof ActionBarActivity){
            ActionBar actionBar = ((ActionBarActivity) getActivity()).getSupportActionBar();
            if(actionBar != null){
                actionBar.setTitle(subtitle);
            }
        }
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        getActivity().getMenuInflater().inflate(R.menu.menu_search_view, menu);
        SearchView searchView = (SearchView) MenuItemCompat.getActionView(menu.findItem(R.id.menu_searchView));
        searchView.setQueryHint(searchKeyword);
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String s) {
                s = s.trim();
                if("".equals(s)){
                    Toast.makeText(getActivity(), "搜索关键字不能为空", Toast.LENGTH_LONG).show();
                    return false;
                }

                setTitle(s);
                Bundle bundle = new Bundle();
                bundle.putString(SearchFragment.PARAM_OPTIONAL_STRING_SEARCH_KEYWORD, s);
                SearchFragment searchFragment = new SearchFragment();
                searchFragment.setArguments(bundle);
                getFragmentManager()
                        .beginTransaction()
                        .setCustomAnimations(R.anim.window_push_enter, R.anim.window_push_exit)
                        .replace(R.id.frame_main_content, searchFragment)
                        .commit();

                ((InputMethodManager) getActivity().getSystemService(Context.INPUT_METHOD_SERVICE))
                        .hideSoftInputFromWindow(getActivity().getCurrentFocus()
                                        .getWindowToken(),
                                InputMethodManager.HIDE_NOT_ALWAYS);

                return true;
            }

            @Override
            public boolean onQueryTextChange(String s) {
                return false;
            }
        });

        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        pullRefreshLayout.setOnRefreshListener(this);

        recyclerView.setOnScrollListener(new ScrollingPauseLoadManager(view.getContext()));

        if (searchImageListAdapter == null) {
            pullRefreshLayout.startRefresh();
        } else {
            recyclerView.setAdapter(searchImageListAdapter);
            recyclerView.scheduleLayoutAnimation();
        }
    }

    @Override
    public void onDetach() {
        if (refreshRequestFuture != null && !refreshRequestFuture.isFinished()) {
            refreshRequestFuture.cancel(true);
        }

        super.onDetach();
    }

    @Override
    public void onRefresh() {
        if (refreshRequestFuture != null && !refreshRequestFuture.isFinished()) {
            return;
        }

        loadMoreListener.cancel();
        searchImageRequest.setStart(0);
        refreshRequestFuture = GoHttp.with(getActivity()).newRequest(searchImageRequest, new JsonHttpResponseHandler(SearchImageRequest.Response.class), new HttpRequest.Listener<SearchImageRequest.Response>() {
            @Override
            public void onStarted(HttpRequest httpRequest) {
                hintView.hidden();
            }

            @Override
            public void onCompleted(HttpRequest httpRequest, HttpResponse httpResponse, SearchImageRequest.Response responseObject, boolean b, boolean b2) {
                if (getActivity() == null) {
                    return;
                }

                List<StarImageRequest.Image> imageList = new ArrayList<StarImageRequest.Image>();
                for (SearchImageRequest.Image image : responseObject.getImages()) {
                    imageList.add(image);
                }

                recyclerView.setAdapter(searchImageListAdapter = new SearchImageAdapter(getActivity(), null, imageList, SearchFragment.this));
                recyclerView.scheduleLayoutAnimation();
                pullRefreshLayout.stopRefresh();
                loadMoreListener.reset();
                searchImageListAdapter.setOnLoadMoreListener(loadMoreListener);
            }

            @Override
            public void onFailed(HttpRequest httpRequest, HttpResponse httpResponse, HttpRequest.Failure failure, boolean b, boolean b2) {
                if (getActivity() == null) {
                    return;
                }

                pullRefreshLayout.stopRefresh();
                if (searchImageListAdapter == null) {
                    hintView.failure(failure, new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            pullRefreshLayout.startRefresh();
                        }
                    });
                } else {
                    Toast.makeText(getActivity(), "刷新失败", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onCanceled(HttpRequest httpRequest) {

            }
        }).responseHandleCompletedAfterListener(new SearchImageRequest.ResponseHandler()).go();
    }

    @Override
    public void onItemClick(int position, StarImageRequest.Image image) {
        DetailActivity.launch(getActivity(), (ArrayList<String>) searchImageListAdapter.getImageUrlList(), position);
    }

    private class MyLoadMoreListener implements SearchImageAdapter.OnLoadMoreListener {
        private boolean end;
        private HttpRequestFuture loadMoreRequestFuture;

        @Override
        public boolean isEnd() {
            return end;
        }

        public void reset(){
            end = false;
        }

        @Override
        public void onLoadMore() {
            if (refreshRequestFuture != null && !refreshRequestFuture.isFinished()) {
                return;
            }

            searchImageRequest.setStart(searchImageListAdapter.getDataSize());
            loadMoreRequestFuture = GoHttp.with(getActivity()).newRequest(searchImageRequest, new JsonHttpResponseHandler(SearchImageRequest.Response.class), new HttpRequest.Listener<SearchImageRequest.Response>() {
                @Override
                public void onStarted(HttpRequest httpRequest) {

                }

                @Override
                public void onCompleted(HttpRequest httpRequest, HttpResponse httpResponse, SearchImageRequest.Response responseObject, boolean b, boolean b2) {
                    if (getActivity() == null) {
                        return;
                    }


                    List<StarImageRequest.Image> newImageList = null;
                    if (responseObject.getImages() != null) {
                        newImageList = new ArrayList<StarImageRequest.Image>();
                        for (SearchImageRequest.Image image : responseObject.getImages()) {
                            newImageList.add(image);
                        }
                    }

                    int count = searchImageListAdapter.getCount();

                    if (newImageList != null && newImageList.size() > 0) {
                        searchImageListAdapter.append(newImageList);
                        if (newImageList.size() < searchImageRequest.getSize()) {
                            end = true;
                            Toast.makeText(getActivity(), "新送达" + newImageList.size() + "个包裹，已全部送完！", Toast.LENGTH_SHORT).show();
                        } else {
                            Toast.makeText(getActivity(), "新送达" + newImageList.size() + "个包裹", Toast.LENGTH_SHORT).show();
                        }
                    } else {
                        end = true;
                        Toast.makeText(getActivity(), "没有您的包裹了", Toast.LENGTH_SHORT).show();
                    }
                    searchImageListAdapter.notifyDataSetChanged();
                }

                @Override
                public void onFailed(HttpRequest httpRequest, HttpResponse httpResponse, HttpRequest.Failure failure, boolean b, boolean b2) {
                    if (getActivity() == null) {
                        return;
                    }
                    searchImageListAdapter.loadMoreFail();
                    Toast.makeText(getActivity(), "快递投递失败", Toast.LENGTH_SHORT).show();
                }

                @Override
                public void onCanceled(HttpRequest httpRequest) {

                }
            }).responseHandleCompletedAfterListener(new SearchImageRequest.ResponseHandler()).go();
        }

        public void cancel() {
            if (loadMoreRequestFuture != null && !loadMoreRequestFuture.isFinished()) {
                loadMoreRequestFuture.cancel(true);
            }
        }
    }
}

File: sample/src/main/java/me/panpf/sketch/sample/ui/SearchFragment.kt
package me.panpf.sketch.sample.ui

import android.content.Context
import android.os.Bundle
import androidx.annotation.NonNull
import androidx.core.view.MenuItemCompat
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import android.view.Menu
import android.view.MenuInflater
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Toast
import com.google.android.flexbox.FlexboxItemDecoration
import com.google.android.flexbox.FlexboxLayoutManager
import kotlinx.android.synthetic.main.fragment_recycler.*
import me.panpf.adapter.AssemblyAdapter
import me.panpf.adapter.AssemblyRecyclerAdapter
import me.panpf.adapter.more.OnLoadMoreListener
import me.panpf.sketch.sample.R
import me.panpf.sketch.sample.base.BaseFragment
import me.panpf.sketch.sample.base.BindContentView
import me.panpf.sketch.sample.bean.BaiduImage
import me.panpf.sketch.sample.bean.BaiduImageSearchResult
import me.panpf.sketch.sample.bean.Image
import me.panpf.sketch.sample.event.ChangeMainPageBgEvent
import me.panpf.sketch.sample.item.LoadMoreItemFactory
import me.panpf.sketch.sample.item.StaggeredImageItemFactory
import me.panpf.sketch.sample.net.NetServices
import me.panpf.sketch.sample.util.ScrollingPauseLoadManager
import me.panpf.sketch.sample.widget.HintView
import org.greenrobot.eventbus.EventBus
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.lang.ref.WeakReference

/**
 * 图片搜索Fragment
 */
@BindContentView(R.layout.fragment_recycler)
class SearchFragment : BaseFragment(), StaggeredImageItemFactory.OnItemClickListener, androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener, OnLoadMoreListener {

    private var searchKeyword: String? = "GIF"

    private var pageIndex = 1
    private var adapter: AssemblyRecyclerAdapter? = null

    private var backgroundImageUri: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)

        val arguments = arguments
        if (arguments != null) {
            searchKeyword = arguments.getString(PARAM_OPTIONAL_STRING_SEARCH_KEYWORD)
            if (searchKeyword == null) {
                searchKeyword = "GIF"
            }
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        setTitle(searchKeyword ?: "")
    }

    private fun setTitle(subtitle: String) {
        if (activity != null && activity is AppCompatActivity) {
            val actionBar = (activity as AppCompatActivity).supportActionBar
            if (actionBar != null) {
                actionBar.subtitle = subtitle
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater!!.inflate(R.menu.menu_search_view, menu)
        @Suppress("DEPRECATION")
        val searchView = MenuItemCompat.getActionView(menu!!.findItem(R.id.menu_searchView)) as SearchView
        searchView.queryHint = searchKeyword
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(s: String): Boolean {
                val activity = activity ?: return false
                val fragmentManager = fragmentManager ?: return false

                val keyword = s.trim { it <= ' ' }
                if ("" == keyword) {
                    Toast.makeText(activity, "搜索关键字不能为空", Toast.LENGTH_LONG).show()
                    return false
                }

                setTitle(keyword)
                val bundle = Bundle()
                bundle.putString(PARAM_OPTIONAL_STRING_SEARCH_KEYWORD, keyword)
                val searchFragment = SearchFragment()
                searchFragment.arguments = bundle
                fragmentManager
                        .beginTransaction()
                        .setCustomAnimations(R.anim.window_push_enter, R.anim.window_push_exit)
                        .replace(R.id.mainFm_contentFrame, searchFragment)
                        .commit()

                (activity.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager)
                        .hideSoftInputFromWindow(activity.currentFocus!!.windowToken, InputMethodManager.HIDE_NOT_ALWAYS)

                return true
            }

            override fun onQueryTextChange(s: String): Boolean {
                return false
            }
        })

        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onViewCreated(@NonNull view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        refresh_recyclerFragment.setOnRefreshListener(this)

        recycler_recyclerFragment_content.addOnScrollListener(ScrollingPauseLoadManager(view.context))

        recycler_recyclerFragment_content.layoutManager = FlexboxLayoutManager(context)
        recycler_recyclerFragment_content.addItemDecoration(FlexboxItemDecoration(context))

        if (adapter == null) {
            refresh_recyclerFragment.post { onRefresh() }
        } else {
            setAdapter(adapter)
        }
    }

    override fun onDestroyView() {
        setTitle("")
        super.onDestroyView()
    }

    override fun onUserVisibleChanged(isVisibleToUser: Boolean) {
        if (isVisibleToUser) {
            changeBackground(backgroundImageUri)
        }
    }

    private fun changeBackground(imageUri: String?) {
        this.backgroundImageUri = imageUri
        backgroundImageUri?.let { EventBus.getDefault().post(ChangeMainPageBgEvent(it)) }
    }

    private fun setAdapter(adapter: AssemblyRecyclerAdapter?) {
        recycler_recyclerFragment_content.adapter = adapter
        recycler_recyclerFragment_content.scheduleLayoutAnimation()
        this.adapter = adapter
    }

    override fun onRefresh() {
        adapter?.loadMoreFinished(false)

        if (!refresh_recyclerFragment.isRefreshing) {
            refresh_recyclerFragment.isRefreshing = true
        }

        loadData(1)
    }

    override fun onItemClick(position: Int, image: BaiduImage, loadingImageOptionsInfo: String) {
        val activity = activity ?: return
        @Suppress("UNCHECKED_CAST")
        val imageList = adapter!!.dataList as List<BaiduImage>
        val urlList = imageList.map { Image(it.url!!, it.url!!) }
        ImageDetailActivity.launch(activity, dataTransferHelper.put("urlList", urlList), loadingImageOptionsInfo, position - adapter!!.headerItemCount)
    }

    private fun loadData(pageIndex: Int) {
        this.pageIndex = pageIndex
        val pageStart = (pageIndex - 1) * PAGE_SIZE
        NetServices.baiduImage().searchPhoto(searchKeyword, searchKeyword, pageStart, PAGE_SIZE).enqueue(LoadDataCallback(this, pageIndex))
    }

    override fun onLoadMore(adapter1: AssemblyAdapter) {
        loadData(pageIndex + 1)
    }

    private class LoadDataCallback internal constructor(fragment: SearchFragment, private val pageIndex: Int) : Callback<BaiduImageSearchResult> {

        private val reference: WeakReference<SearchFragment> = WeakReference(fragment)

        init {
            if (pageIndex == 1) {
                fragment.hint_recyclerFragment.hidden()
            }
        }

        override fun onResponse(call: Call<BaiduImageSearchResult>, response: Response<BaiduImageSearchResult>) {
            val fragment = reference.get() ?: return
            if (!fragment.isViewCreated) {
                return
            }

            filterEmptyImage(response)

            if (pageIndex == 1) {
                create(fragment, response)
            } else {
                loadMore(fragment, response)
            }

            fragment.refresh_recyclerFragment.isRefreshing = false
        }

        override fun onFailure(call: Call<BaiduImageSearchResult>, t: Throwable) {
            val fragment = reference.get() ?: return
            val activity = fragment.activity ?: return
            if (!fragment.isViewCreated) {
                return
            }

            if (pageIndex == 1) {
                fragment.hint_recyclerFragment.failed(t, View.OnClickListener {
                    fragment.onRefresh()
                })
                fragment.refresh_recyclerFragment.isRefreshing = false
            } else {
                fragment.adapter!!.loadMoreFailed()
                Toast.makeText(fragment.activity, HintView.getCauseByException(activity, t), Toast.LENGTH_LONG).show()
            }
        }

        private fun filterEmptyImage(response: Response<BaiduImageSearchResult>) {
            val imageList = response.body()!!.imageList ?: return

            val mutableImageList = imageList.toMutableList()
            val imageIterator = mutableImageList.iterator()
            while (imageIterator.hasNext()) {
                val image = imageIterator.next()
                if (image.url == null || "" == image.url) {
                    imageIterator.remove()
                }
            }
            response.body()!!.imageList = mutableImageList
        }

        private fun create(fragment: SearchFragment, response: Response<BaiduImageSearchResult>) {

            val images = response.body()!!.imageList
            if (images == null || images.size == 0) {
                fragment.hint_recyclerFragment.empty("No photos")
                return
            }

            val adapter = AssemblyRecyclerAdapter(images)
            adapter.addItemFactory(StaggeredImageItemFactory(fragment))
            adapter.setMoreItem(LoadMoreItemFactory(fragment).fullSpan(fragment.recycler_recyclerFragment_content))

            fragment.recycler_recyclerFragment_content.adapter = adapter
            fragment.adapter = adapter

            fragment.changeBackground(images[0].url)
        }

        private fun loadMore(fragment: SearchFragment, response: Response<BaiduImageSearchResult>) {

            val images = response.body()!!.imageList
            if (images == null || images.size == 0) {
                fragment.adapter!!.loadMoreFinished(true)
                return
            }

            fragment.adapter!!.addAll(images)
            fragment.adapter!!.loadMoreFinished(images.size < 20)

            fragment.changeBackground(images[0].url)
        }
    }

    companion object {
        val PARAM_OPTIONAL_STRING_SEARCH_KEYWORD = "PARAM_OPTIONAL_STRING_SEARCH_KEYWORD"
        private val PAGE_SIZE = 60
    }
}
